File: src/transform.ts
import { Matrix3, Matrix4 } from "./matrix";
import { Vector2D, Vector3D } from "./vector";

export class Transform2D{
    _scale: Vector2D = new Vector2D(1,1)
    _translation: Vector2D = new Vector2D(0,0)
    changed: Boolean = false
    cached: Matrix3

    translate(translation: Vector2D): Transform2D{
        this.changed = true
        this._translation = this._translation.translate(translation)
        return this
    }
    scale(scale: Vector2D): Transform2D{
        this.changed = true
        this._scale = this._scale.scaleV(scale)
        return this
    }
    transform(): Matrix3{
        if(!this.cached || this.changed){
            this.cached = Matrix3.identity().scale(this._scale).translate(this._translation)
        }
        return this.cached
    }
}
export class Transform3D{
    _scale: Vector3D = new Vector3D(1,1,1)
    _translation: Vector3D = new Vector3D(0,0,0)
    changed: Boolean = false
    cached: Matrix4

    translate(translation: Vector3D): Transform3D{
        this.changed = true
        this._translation = this._translation.translate(translation)
        return this
    }
    scale(scale: Vector3D): Transform3D{
        this.changed = true
        this._scale = this._scale.scaleV(scale)
        return this
    }
    transform(): Matrix4{
        if(!this.cached || this.changed){
            this.cached = Matrix4.identity().scale(this._scale).translate(this._translation)
        }
        return this.cached
    }
}

File: src/sgl/Transform.kt
package sgl

interface Transformable{
    @JsName("transform") fun transform(vector: Vector3): Vector3
    @JsName("translate") fun translate(translation: Vector3)
    @JsName("rotate") fun rotate(angle: Float, axis: Vector3)
    @JsName("rotateAbout") fun rotate(about: Vector3, angle: Float, axis: Vector3)
    @JsName("scale") fun scale(scale: Vector3)
}

open class Transform: Transformable{
    var matrix: Matrix4 = Matrix4.identity
        get() {
            if (dirty) {
                matrix = Matrix4.translation(translationVector) * Matrix4.scale(scaleVector) * rotation.transformation() * transformations
                dirty = false
            }
            return field
        }
    var dirty = true
    var translationVector = Vector3.zero
        set(value) { dirty = true; field = value }
    var scaleVector = vectorOf(1f,1f,1f)
        set(value) { dirty = true; field = value }
    var rotation = Quaternion.from(0f, vectorOf(1f,0f,0f))
        set(value) { dirty = true; field = value }
    var transformations = Matrix4.identity
        set(value) { dirty = true; field = value }

    override fun translate(translation: Vector3){
        translationVector += translation
    }
    override fun rotate(angle: Float, axis: Vector3) {
        rotation *= Quaternion.from(angle, axis)
    }
    override fun rotate(about: Vector3, angle: Float, axis: Vector3) {
        val localPosition = translationVector + about
        transformations *= Matrix4.translation(localPosition*-1f) * Quaternion.from(angle, axis).transformation() * Matrix4.translation(localPosition)
    }
    override fun scale(scale: Vector3){
        scaleVector *= scale
    }

    override fun transform(vector: Vector3) = matrix * vector
    operator fun invoke(vector: Vector3) = transform(vector)
}

open class InverseTransform: Transform(){
    var inverseMatrix: Matrix4 = matrix.inverse()
        get() = matrix.inverse()

    override fun transform(vector: Vector3) = inverseMatrix * vector
}